### ISSUE TITLE
Managing TLS certificates

### ISSUE THREAD

**Issue by nathany** [convox/rack#212]
What is the possibility of integrating ACME to handle SSL certificates, using a service like the free [Let's Encrypt](https://letsencrypt.org)? I'm thinking both for the Convox API rather than a self-signed cert, and for customers as an alternative to purchasing and managing certificates for applications.

There is an article going up on the [Gopher Academy](https://blog.gopheracademy.com/) blog on Tuesday about using [lego](https://github.com/xenolf/lego) for the ACME protocol ([early access](https://github.com/gopheracademy/gopheracademy-web/pull/155/files)).

What's not clear to me is if the current ACME challenges will work with ELB? There's also a matter of managing certificate renewals automatically once installed.

**Comment by csquared** [comment]
We've been tracking Let's Encrypt for a while now and are just figuring out when/how, not if, to cut over. Now's the time to make suggestions!

**Comment by csquared** [comment]
(PRs welcome as well)

**Comment by csquared** [comment]
Thanks for the early access link to the article!

**Comment by nathany** [comment]
The article is live now: https://blog.gopheracademy.com/advent-2015/generate-free-tls-certs-with-lego/

**Comment by mholt** [comment]
FYI, the DNS challenge was recently turned on for LE's staging endpoint, so you can experiment with that if you need to.

**Comment by nathany** [comment]
This article uses the ACME Python client to secure Nginx in a Docker container:
https://getcarina.com/blog/push-button-lets-encrypt/

**Comment by slaskis** [comment]
Another implementation that I think seems very promising is https://github.com/hlandau/acme

**Comment by alex** [comment]
I've been playing with an implementation using DNS challenges + ELB. Not usable yet, but maybe a useful model: github.com/alex/letsencrypt-aws/

**Comment by csquared** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;38SjOutM&#x2F;259-generate-ssl-certs-with-lets-encrypt"><img src="https:&#x2F;&#x2F;github.trello.services&#x2F;images&#x2F;trello-icon.png" width="12" height="12"> generate ssl certs with lets encrypt</a>

**Comment by nzoschke** [comment]
@mattmanning can you update this issue with your findings about integrating Lets Encrypt?

**Comment by alex** [comment]
FWIW, I suspect the answer is "not easy, unless we control DNS". The other
two challenges are SNI and HTTP. ELBs do not offer SNI, and AFAIK convox
goes directly from ELB -> your application, so there's no place to sneak
into that path and control the challenge.

On Mon, Jan 18, 2016 at 8:50 PM, Noah Zoschke <notifications@github.com>
wrote:

> @mattmanning <https://github.com/mattmanning> can you update this issue
> with your findings about integrating Lets Encrypt?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/convox/rack/issues/212#issuecomment-172704450>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

**Comment by nathany** [comment]
Maybe it could be used to secure the Convox API as a start, but not your actual app?

**Comment by nathany** [comment]
@bgentry pointed out this in Slack. Sounds like it achieves the same ends.

https://aws.amazon.com/blogs/aws/new-aws-certificate-manager-deploy-ssltls-based-apps-on-aws/

**Comment by cozmo** [comment]
Unfortunately AWS cert manager only works in US East right now. It also relies on email for domain validation, which makes it difficult to use in an automated fashion (since a user has to check their email and click a link before the cert generation can continue). That said I think if it can be made to work in US East it's probably the best tool for the job. 

I've been using https://github.com/alex/letsencrypt-aws to setup letsencrypt on all my public ELBs. It requires no manual intervention since it uses route53 to verify domain ownership (via TXT records). I just have a dockerfile running it in the background making sure my certs are up to date. I think a workflow like that would work very well if built into `convox ssl`. You could basically `convox ssl setup --app myapp --domain mydomain.com` and convox could run the same steps as letsencrypt-aws does. 

I would be very excited for this and would be happy to contribute if it was a direction convox was interested in going.

**Comment by mwarkentin** [comment]
Had a discussion with @ddollar in slack the other day about the various ways that SSL certs could be handled. 

Use cases:

* Existing cert from external provider (private/public key to upload to IAM)
* Existing cert in AWS IAM or ACM (ARN identifier)
* New cert required (convox can generate it)
* Single domain or wildcard certs shared across ELBs

There are some concerns with the existing `convox ssl create` command becoming to overloaded with options if we were to try to use it to handle all of these use cases as you can see:

```
$ convox ssl create --help
convox ssl create: create a new SSL listener

Usage:
  convox ssl create <process:port>

Options:
   --app, -a         App name. Inferred from current directory if not specified.
   --arn    Reference to an existing certificate from IAM or ACM
   --chain         Intermediate certificate chain.
   --domain   Domain to generate the ACM certificate for
   --private-key  Path to your SSL cert private key. Requires --public-key to be set.
   --public-key   Path to your SSL cert public key. Requires --private-key to be set.
   --secure        Use a TLS-encrypted listener.
   --self-signed    Generate a self-signed cert.
   --verifying-domain   Domain used to verify the ACM certificate. Must be a superset of --domain.
```

It might make sense to come up with an alternative abstraction in order to simplify the management of certs while satisfying the above use cases. One option could be to move management of certs into a service, and change `convox ssl` to simply take an ARN as an argument to create the listener. I've broken up the options by "certificate type" - not sure if this helps keep things better organized or not.

An example of how that could look:

```
$ convox services create cert --help
convox services create cert: create a new SSL certificate

Usage:
  convox services create cert

Options:
   --acm-domain   Domain to generate the ACM certificate for
   --acm-verifying-domain   Domain used to verify the ACM certificate. Must be a superset of --domain.
   --iam-chain         Intermediate certificate chain.
   --iam-private-key  Path to your SSL cert private key. Requires --public-key to be set.
   --iam-public-key   Path to your SSL cert public key. Requires --private-key to be set.
   --iam-self-signed    Generate a self-signed cert and store in IAM.
```

Once we had the `cert` service, creating an SSL listener would be as simple as:

```
$ convox ssl create --help
convox ssl create: create a new SSL listener

Usage:
  convox ssl create <process:port>

Options:
   --app, -a         App name. Inferred from current directory if not specified.
   --arn    Reference to an existing certificate from IAM or ACM
   --secure        Use a TLS-encrypted listener.
```

This would also allow you to use existing certs which are not managed by Convox. In the meantime, our workaround was to create a self-signed cert with convox, and then replace it on the listener with our ACM cert manually.

**Comment by mwarkentin** [comment]
Adding the `--arn` option as a first step would allow people to use ACM, etc. without having to use the `--self-signed` workaround - larger changes like a `cert` service could then be handled later.

**Comment by ddollar** [comment]
I'm :+1: for adding an `--arn` option.

**Comment by ddollar** [comment]
This should all be addressed by #563 and #568. Thanks for the feedback!
